subsys/caf/modules/settings_loader.c

```diff
@@ -65,55 +67,35 @@ static void load_settings(void)
 
 static bool module_event_handler(const struct module_state_event *event)
 {
-	/* Settings need to be loaded after all client modules are ready. */
-	const void * const req_modules[] = {
-		MODULE_ID(main),
-#if CONFIG_DESKTOP_HIDS_ENABLE
-		MODULE_ID(hids),
-#endif
-#if CONFIG_DESKTOP_BAS_ENABLE
-		MODULE_ID(bas),
-#endif
-#if CONFIG_CAF_BLE_ADV
-		MODULE_ID(ble_adv),
-#endif
-#if CONFIG_DESKTOP_MOTION_SENSOR_ENABLE
-		MODULE_ID(motion),
-#endif
-#if CONFIG_DESKTOP_BLE_SCANNING_ENABLE
-		MODULE_ID(ble_scan),
-#endif
-#if CONFIG_DESKTOP_FAILSAFE_ENABLE
-		MODULE_ID(failsafe),
-#endif
-	};
-
-	static uint32_t req_state;
-
-	BUILD_ASSERT(ARRAY_SIZE(req_modules) < (8 * sizeof(req_state)),
-			"Array size bigger than number of bits");
-
-	if (req_state == BIT_MASK(ARRAY_SIZE(req_modules))) {
-		/* Already initialized */
-		return false;
-	}
+	static struct module_flags req_modules_bm;
+
+	if (check_state(event, MODULE_ID(main), MODULE_STATE_READY)) {
+		static bool initialized;
+
+		__ASSERT_NO_MSG(!initialized);
+		initialized = true;
 
-	for (size_t i = 0; i < ARRAY_SIZE(req_modules); i++) {
-		if (check_state(event, req_modules[i], MODULE_STATE_READY)) {
-			unsigned int flag = BIT(i);
+		get_req_modules(&req_modules_bm);
 
-			/* Catch double initialization of any module */
-			__ASSERT_NO_MSG((req_state & flag) == 0);
+		/* In case user implemented empty get_req_modules function */
+		if (module_flags_check_zero(&req_modules_bm)) {
+			load_settings();
+		}
+	}
 
-			req_state |= flag;
+	if (module_flags_check_zero(&req_modules_bm)) {
```

Comment:
I would put this at the very beginning of `module_state_event` handler to reduce number of operations when e.g. system goes to power down state